filters/include/pcl/filters/impl/random_sample.hpp

```diff
@@ -112,37 +112,36 @@ pcl::RandomSample<PointT>::applyFilter (std::vector<int> &indices)
     // Set random seed so derived indices are the same each time the filter runs
     std::srand (seed_);
 
-    // Algorithm A
-    unsigned top = N - sample_size;
-    unsigned i = 0;
-    unsigned index = 0;
+    // Algorithm S
+    unsigned int i = 0;
+    unsigned int index = 0;
     std::vector<bool> added;
     if (extract_removed_indices_)
       added.resize (indices_->size (), false);
-    for (size_t n = sample_size; n >= 2; n--)
+    size_t n = sample_size;
+    while (n > 0)
     {
-      float V = unifRand ();
-      unsigned S = 0;
-      float quot = static_cast<float> (top) / static_cast<float> (N);
-      while (quot > V)
+      // Step 1: [Generate U.] Generate a random variate U that is uniformly distributed between 0 and 1.
+      const float U = unifRand ();
+      // Step 2: [Test.] If N * U > n, go to Step 4. 
+      if ((N * U) <= n)
       {
-        S++;
-        top--;
-        N--;
-        quot = quot * static_cast<float> (top) / static_cast<float> (N);
+        // Step 3: [Select.] Select the next record in the file for the sample, and set n : = n - 1 and N : = N - 1.
+        // If n > 0, then return to Step 1; otherwise, the sample is complete and the algorithm terminates.
+        if (extract_removed_indices_)
+          added[index] = true;
+        indices[i++] = (*indices_)[index++];
+        --n;
+        --N;
+      }
+      else
+      {
+        // Step 4: [Don't select.] Skip over the next record (do not include it in the sample), set N : = N - 1, and return to Step 1.
+        ++index;
+        --N;
```

Comment:
Regardless of U, you always decrement N and you always increment index. I suggest you move these operations outside of the if/else block. Basically the else is unnecessary.

Reply:
How about an empty "else" which is removed by the compiler for better readability?
```
while (n > 0)
{
  // Step 1: [Generate U.] Generate a random variate U that is uniformly distributed between 0 and 1.
  const float U = unifRand ();
  // Step 2: [Test.] If N * U > n, go to Step 4. 
  if ((N * U) <= n)
  {
    // Step 3: [Select.] Select the next record in the file for the sample, and set n : = n - 1.
    if (extract_removed_indices_)
      added[index] = true;
    indices[i++] = (*indices_)[index];
    --n;
  }
  else
  {
    // Step 4: [Don't select.] Skip over the next record (do not include it in the sample).
  }
  // Set N : = N - 1.
  --N;
  ++index;
  // If n > 0, then return to Step 1; otherwise, the sample is complete and the algorithm terminates.
}
```

Reply:
I wouldn't include it as actual code, but I'm ok with a commented version of the else in order to visualize the "step 4" in context.